Add mouse button bindings

coloursofnoise opened the pull request:
Add support for binding mouse buttons to inputs.
Mouse buttons are set in the Keyboard config menu, and Keys take priority wherever necessary.

Still missing are icons for the buttons to display in the config menus [[1]](https://github.com/EverestAPI/Everest/blob/f05acb8d90d4738462d1c88f2b26023fe877a44d/Celeste.Mod.mm/Patches/TextMenu.cs#L329)

coloursofnoise committed:
Add Mouse to Binding and MInput classes

coloursofnoise committed:
Display Mouse Bindings in TextMenu

coloursofnoise committed:
Add mouse remapping to vanilla KeyboardConfigUI

coloursofnoise committed:
Save and Load MouseButton bindings separately from vanilla binds

coloursofnoise committed:
Add MouseButtons to Everest ButtonBinding

coloursofnoise committed:
Merge branch 'drop-legacy-support' of https://github.com/coloursofnoise/Everest into mousebuttonbinding

coloursofnoise commented:
This PR is on hold pending the merge of 
- #327

coloursofnoise committed:
Merge branch 'dev' of https://github.com/EverestAPI/Everest into mousebuttonbinding

WEGFan commented:
I recently watched [this video](https://youtu.be/d6GtGbI-now) and there's a [button prompt pack](https://thoseawesomeguys.com/prompts/) in it, can it fit our needs for the config menu?

coloursofnoise committed:
Add icons for Mouse ButtonBindings

coloursofnoise committed:
Fix crash when mousebindings file not present

coloursofnoise committed:
Ensure keyboard bindings aren't reset if mouse bindings present

coloursofnoise committed:
Merge branch 'dev' into mousebuttonbinding

coloursofnoise committed:
Refactor TextMenu.Setting.Update foreach patch into separate method

It wasn't worth it.

coloursofnoise committed:
Add documentation and fix names to align with new functions

Kalobi commented on Celeste.Mod.mm/MonoModRules.cs:
```diff
@@ -339,9 +339,10 @@ class PatchAreaCompleteVersionNumberAndVariantsAttribute : Attribute { };
 
     /// <summary>
     /// Patches {Button,Keyboard}ConfigUI.Update (InputV2) to call a new Reset method instead of the vanilla one.
+    /// Also implments mouse button remapping.
```
```suggestion
    /// Also implements mouse button remapping.
```

bigkahuna443 commented on Celeste.Mod.mm/Patches/Settings.cs:
```diff
@@ -157,5 +159,45 @@ class patch_Settings : Settings {
+            }
+        }
+
+        public void ClearMouseControls() {
```
I feel like this would be a better fit as a method in `VanillaMouseBindings` that you call from here.

coloursofnoise replied:
This matches the vanilla `Settings.SetDefault{Keyboard/Button}Controls`, and I would prefer to keep the scope for `VanillaMouseBindings` restricted to a serialization helper.

bigkahuna443 commented on Celeste.Mod.mm/Patches/OuiOptions.cs:
```diff
@@ -0,0 +1,21 @@
+namespace Celeste {
+    class patch_OuiOptions : OuiOptions {
+        public extern IEnumerator orig_Leave(Oui next);
+        public override IEnumerator Leave(Oui next) {
```
Re: discussion in Discord - this only gets called from the main menu options. If you want settings to apply after changes from the in-game menu, you'd probably want to add this in `UserIO` instead

coloursofnoise committed:
Move mousebindings save to UserIO.SaveHandler

Ensures that bindings are saved whenever settings are,
not just when exiting OuiOptions

bigkahuna443 commented on Celeste.Mod.mm/Patches/UserIO.cs:
```diff
@@ -54,6 +55,13 @@ public static string GetSaveFilePath(string name = null)
+                UserIO.Save<VanillaMouseBindings>("modsettings-Everest_MouseBindings", data);
+                UserIO.Close();
+            }
+
```
If you save settings here it bypasses the save thread and any error handling. Might make more sense to add the serialization in `SaveRoutine` and do the saving in `orig_SaveThread` next to where it does that for vanilla settings?

coloursofnoise replied:
Serializing and Saving have been separated, with saving happening in the `SaveThread`, and mirrors how Mod SaveData is handled.

coloursofnoise committed:
Use ILCursor for SaveRoutine patch

coloursofnoise committed:
Serialize and save mouse bindings separately while using SaveThread

bigkahuna443 commented:
Tested new version, binds work as expected, and they save and load correctly from in-game menu and overworld menu.

Re: Discord discussion, I think it would be better if names were changed to better distinguish from EverestModule settings since mouse bindings are more extensions of vanilla settings (and also mod settings are saved independently of everything else).

Alternatively, the mouse bindings could be bundled into the CoreModule settings, but this would require shuffling some logic around.

coloursofnoise committed:
Rename mousebindings members to clarify purpose
